File: 2390-naming-a-company/2390-naming-a-company.py
class Solution:
    def distinctNames(self, ideas: List[str]) -> int:
        groups = defaultdict(set)

        for idea in ideas:
            suf, pre = idea[1:], idea[0]
            groups[pre].add(suf)
            
        
        ans = 0

        for group1 in groups:
            for group2 in groups:
                if group1 == group2: continue

                ans += len(list(groups[group1] - groups[group2])) * len(list(groups[group2] - groups[group1]))

        return ans

File: 2390-naming-a-company/2390-naming-a-company.cpp
class Solution {
public:
    long long distinctNames(vector<string>& ideas) {
        vector<unordered_set<string>> groups(26);

        for (int i = 0; i < ideas.size(); i++) {
            char prefix = ideas[i][0];
            string suffix = ideas[i].substr(1);
            
            groups[prefix - 'a'].insert(suffix);
        }

        long ans = 0;
        for (int i = 0; i < 26; i++) {
            for (int j = i + 1; j < 26; j++) {
                long count1 = 0, count2 = 0;

                for (auto const suffix : groups[i]) {
                    if (groups[j].find(suffix) == groups[j].end()) {
                        count1++;
                    }
                }

                for (auto const suffix : groups[j]) {
                    if (groups[i].find(suffix) == groups[i].end()) {
                        count2++;
                    }
                }

                ans += count1 * count2;
            }
        }

        return ans * 2;
    }
};